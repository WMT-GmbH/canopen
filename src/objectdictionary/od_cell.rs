pub struct OdCell<T> {
    pub value: T,
    pub locked: bool,
}

impl<T> OdCell<T> {
    pub fn new(value: T) -> Self {
        Self {
            value,
            locked: false,
        }
    }
    pub fn is_locked(&self) -> bool {
        self.locked
    }
    // TODO docs about partial write going on
    pub fn get(&self) -> &T {
        &self.value
    }
    pub fn get_mut(&mut self) -> &mut T {
        self.locked = false;
        &mut self.value
    }
    pub(crate) fn get_mut_unchecked(&mut self) -> &mut T {
        &mut self.value
    }
    pub(crate) fn lock(&mut self) {
        self.locked = true;
    }
    pub(crate) fn unlock(&mut self) {
        self.locked = false;
    }
}
