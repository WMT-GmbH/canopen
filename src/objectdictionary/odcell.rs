use core::cell::{Cell, UnsafeCell};
use core::fmt::Debug;
use core::hint::unreachable_unchecked;
use core::ops::{Deref, DerefMut};

#[derive(Copy, Clone)]
enum BorrowState {
    Unused,
    Locked,
    Reading { locked: bool },
    Writing,
}

/// This is a reimplementation of [`RefCell`](core::cell::RefCell) with an additional flag
/// that tracks if mutable access was granted while a `canopen` service tried
/// to lock the contained value.
///
/// Compared to `RefCell` there are two notable differences:
/// * only one [`Ref`] can be active at a time (consecutive calls to [`borrow`](ODCell::borrow) will panic)
/// * a `locked` flag indicates if a `canopen` service expects the contained value not to change because it is in the middle of a multi-step read/write process
///
/// The state of this `locked` flag can be queried with the [`is_locked`](ODCell::is_locked) method.
///
/// Acquiring mutable access with [`try_borrow_mut`](ODCell::try_borrow_mut) or [`borrow_mut`](ODCell::borrow_mut)
/// will forcibly unlock the Cell.
/// If a canopen service was relying on the variable not being changed,
/// it will see the locked flag not being set and abort the operation it was doing.
pub struct ODCell<T: ?Sized> {
    borrow: Cell<BorrowState>,
    value: UnsafeCell<T>,
}

#[derive(Debug)]
pub struct BorrowError {}

impl<T> ODCell<T> {
    #[inline]
    pub const fn new(value: T) -> ODCell<T> {
        ODCell {
            value: UnsafeCell::new(value),
            borrow: Cell::new(BorrowState::Unused),
        }
    }

    #[inline]
    pub fn into_inner(self) -> T {
        // Since this function takes `self` (the `RefCell`) by value, the
        // compiler statically verifies that it is not currently borrowed.
        self.value.into_inner()
    }
}

impl<T: ?Sized> ODCell<T> {
    #[inline]
    #[track_caller]
    pub fn borrow(&self) -> Ref<'_, T> {
        self.try_borrow().expect("already mutably borrowed")
    }

    #[inline]
    pub fn try_borrow(&self) -> Result<Ref<'_, T>, BorrowError> {
        match BorrowRef::new(&self.borrow) {
            Some(b) => {
                // SAFETY: `BorrowRef` ensures that there is only immutable access
                // to the value while borrowed.
                Ok(Ref {
                    value: unsafe { &*self.value.get() },
                    _borrow: b,
                })
            }
            None => Err(BorrowError {}),
        }
    }

    #[inline]
    #[track_caller]
    pub fn borrow_mut(&self) -> RefMut<'_, T> {
        self.try_borrow_mut().expect("already borrowed")
    }

    #[inline]
    pub fn try_borrow_mut(&self) -> Result<RefMut<'_, T>, BorrowError> {
        match BorrowRefMut::new(&self.borrow) {
            Some(b) => {
                // SAFETY: `BorrowRefMut` guarantees unique access.
                Ok(RefMut {
                    value: unsafe { &mut *self.value.get() },
                    _borrow: b,
                })
            }
            None => Err(BorrowError {}),
        }
    }

    #[inline]
    pub fn get_mut(&mut self) -> &mut T {
        self.value.get_mut()
    }

    #[inline]
    pub fn is_locked(&self) -> bool {
        matches!(
            self.borrow.get(),
            BorrowState::Locked | BorrowState::Reading { locked: true }
        )
    }
}

unsafe impl<T: ?Sized> Send for ODCell<T> where T: Send {}

struct BorrowRef<'b> {
    borrow: &'b Cell<BorrowState>,
}

impl<'b> BorrowRef<'b> {
    #[inline]
    fn new(borrow: &'b Cell<BorrowState>) -> Option<BorrowRef<'b>> {
        match borrow.get() {
            BorrowState::Unused => {
                borrow.set(BorrowState::Reading { locked: false });
                Some(BorrowRef { borrow })
            }
            BorrowState::Locked => {
                borrow.set(BorrowState::Reading { locked: true });
                Some(BorrowRef { borrow })
            }
            _ => None,
        }
    }
}

impl Drop for BorrowRef<'_> {
    #[inline]
    fn drop(&mut self) {
        match self.borrow.get() {
            BorrowState::Reading { locked } => {
                if locked {
                    self.borrow.set(BorrowState::Locked);
                } else {
                    self.borrow.set(BorrowState::Unused);
                }
            }
            _ => {
                debug_assert!(false);
                unsafe { unreachable_unchecked() }
            }
        }
    }
}

pub struct Ref<'b, T: ?Sized> {
    value: &'b T,
    _borrow: BorrowRef<'b>,
}

impl<T: ?Sized> Ref<'_, T> {
    pub(crate) fn lock(self) {
        // this produces the same result as dropping self with the locked flag always set
        debug_assert!(matches!(
            self._borrow.borrow.get(),
            BorrowState::Reading { .. }
        ));
        self._borrow.borrow.set(BorrowState::Locked);
        core::mem::forget(self);
    }
    pub(crate) fn unlock(self) {
        // this produces the same result as dropping self with the locked flag always unset
        debug_assert!(matches!(
            self._borrow.borrow.get(),
            BorrowState::Reading { .. }
        ));
        self._borrow.borrow.set(BorrowState::Unused);
        core::mem::forget(self);
    }
}

impl<T: ?Sized> Deref for Ref<'_, T> {
    type Target = T;

    #[inline]
    fn deref(&self) -> &T {
        self.value
    }
}

struct BorrowRefMut<'b> {
    borrow: &'b Cell<BorrowState>,
}

impl Drop for BorrowRefMut<'_> {
    #[inline]
    fn drop(&mut self) {
        let borrow = self.borrow.get();
        debug_assert!(matches!(borrow, BorrowState::Writing));
        self.borrow.set(BorrowState::Unused);
    }
}

impl<'b> BorrowRefMut<'b> {
    #[inline]
    fn new(borrow: &'b Cell<BorrowState>) -> Option<BorrowRefMut<'b>> {
        match borrow.get() {
            BorrowState::Unused | BorrowState::Locked => {
                borrow.set(BorrowState::Writing);
                Some(BorrowRefMut { borrow })
            }
            _ => None,
        }
    }
}

pub struct RefMut<'b, T: ?Sized> {
    value: &'b mut T,
    _borrow: BorrowRefMut<'b>,
}

impl<T: ?Sized> RefMut<'_, T> {
    pub(crate) fn lock(self) {
        // this produces the same result as dropping self
        // and then switching from BorrowState::Unused to BorrowState::Locked
        debug_assert!(matches!(self._borrow.borrow.get(), BorrowState::Writing));
        self._borrow.borrow.set(BorrowState::Locked);
        core::mem::forget(self);
    }
}

impl<T: ?Sized> Deref for RefMut<'_, T> {
    type Target = T;

    #[inline]
    fn deref(&self) -> &T {
        self.value
    }
}

impl<T: ?Sized> DerefMut for RefMut<'_, T> {
    #[inline]
    fn deref_mut(&mut self) -> &mut T {
        self.value
    }
}
