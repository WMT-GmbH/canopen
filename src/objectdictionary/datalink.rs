use core::cmp::Ordering;
use core::ops::Deref;

use crate::objectdictionary::od_cell::OdCell;
use crate::objectdictionary::{ODError, OdInfo};

mod private {
    use super::{BasicData, OdCell};

    pub trait Sealed {}
    impl<T: BasicData> Sealed for T {}
    impl Sealed for &str {}
    impl Sealed for &[u8] {}
    impl<T> Sealed for OdCell<T> {}
}
/* TODO wait for stabilization
#[diagnostic::on_unimplemented(
    note = "`{Self}` must either implement `BasicData` or be wrapped in `OdCell`"
)]*/
pub trait DataLink: private::Sealed {
    fn read(&mut self, index: u16, subindex: u8) -> Result<ReadData, ODError>;
    fn write(&mut self, data: &WriteData, od_info: OdInfo) -> Result<(), ODError>;
}

pub trait BasicData {
    fn read(&mut self, index: u16, subindex: u8) -> Result<BasicReadData, ODError>;
    fn write(&mut self, data: BasicWriteData, od_info: OdInfo) -> Result<(), ODError>;
}

pub trait CustomData {
    fn read(&self, index: u16, subindex: u8) -> Result<ReadData, ODError>;
    fn write(&mut self, data: WriteStream, od_info: OdInfo) -> Result<(), ODError>;
}

impl<T: BasicData> DataLink for T {
    fn read(&mut self, index: u16, subindex: u8) -> Result<ReadData, ODError> {
        Ok(BasicData::read(self, index, subindex)?.into())
    }

    fn write(&mut self, data: &WriteData, od_info: OdInfo) -> Result<(), ODError> {
        self.write(BasicWriteData(data), od_info)
    }
}

macro_rules! basic_data {
    ($typ:ty) => {
        impl BasicData for $typ {
            fn read(&mut self, _: u16, _: u8) -> Result<BasicReadData, ODError> {
                Ok(BasicReadData::from(*self))
            }

            fn write(&mut self, data: BasicWriteData, _: OdInfo) -> Result<(), ODError> {
                *self = data.try_into()?;
                Ok(())
            }
        }
    };
}

basic_data!(bool);
basic_data!(i8);
basic_data!(i16);
basic_data!(i32);
basic_data!(u8);
basic_data!(u16);
basic_data!(u32);
basic_data!(f32);

impl DataLink for &str {
    fn read(&mut self, _: u16, _: u8) -> Result<ReadData, ODError> {
        Ok(ReadData::Bytes(self.as_bytes()))
    }

    fn write(&mut self, _: &WriteData, _: OdInfo) -> Result<(), ODError> {
        Err(ODError::ReadOnlyError)
    }
}

impl DataLink for &[u8] {
    fn read(&mut self, _: u16, _: u8) -> Result<ReadData, ODError> {
        Ok(ReadData::Bytes(self))
    }

    fn write(&mut self, _: &WriteData, _: OdInfo) -> Result<(), ODError> {
        Err(ODError::ReadOnlyError)
    }
}

impl<T: CustomData> DataLink for OdCell<T> {
    fn read(&mut self, index: u16, subindex: u8) -> Result<ReadData, ODError> {
        // TODO locking
        CustomData::read(self.get(), index, subindex)
    }

    fn write(&mut self, data: &WriteData, od_info: OdInfo) -> Result<(), ODError> {
        if data.is_first_segment() {
            self.lock();
        } else if !self.is_locked() {
            return Err(ODError::LocalControlError);
        }
        if data.is_last_segment {
            self.unlock();
        }
        self.get_mut_unchecked().write(WriteStream(data), od_info)
    }
}

impl<const N: usize> CustomData for [u8; N] {
    fn read(&self, _: u16, _: u8) -> Result<ReadData, ODError> {
        Ok(self[..].into())
    }

    fn write(&mut self, data: WriteStream, _: OdInfo) -> Result<(), ODError> {
        data.write_into(self)?;
        Ok(())
    }
}

pub enum ReadData<'a> {
    B1([u8; 1]),
    B2([u8; 2]),
    B4([u8; 4]),
    Bytes(&'a [u8]),
}

pub enum BasicReadData {
    B1([u8; 1]),
    B2([u8; 2]),
    B4([u8; 4]),
}

impl From<BasicReadData> for ReadData<'_> {
    fn from(value: BasicReadData) -> Self {
        match value {
            BasicReadData::B1(val) => ReadData::B1(val),
            BasicReadData::B2(val) => ReadData::B2(val),
            BasicReadData::B4(val) => ReadData::B4(val),
        }
    }
}

impl ReadData<'_> {
    pub fn as_bytes(&self) -> &[u8] {
        match self {
            ReadData::B1(val) => val,
            ReadData::B2(val) => val,
            ReadData::B4(val) => val,
            ReadData::Bytes(val) => val,
        }
    }
}

impl<'a> From<&'a [u8]> for ReadData<'a> {
    #[inline]
    fn from(val: &'a [u8]) -> Self {
        ReadData::Bytes(val)
    }
}

impl<'a> From<&'a str> for ReadData<'a> {
    #[inline]
    fn from(val: &'a str) -> Self {
        ReadData::Bytes(val.as_bytes())
    }
}

macro_rules! read_data_from_impl {
    ($typ:ty, $variant:path, $basic_variant:path) => {
        impl From<$typ> for ReadData<'_> {
            #[inline]
            fn from(val: $typ) -> Self {
                $variant(val.to_le_bytes())
            }
        }
        impl From<$typ> for BasicReadData {
            #[inline]
            fn from(val: $typ) -> Self {
                $basic_variant(val.to_le_bytes())
            }
        }
    };
}

read_data_from_impl!(u8, ReadData::B1, BasicReadData::B1);
read_data_from_impl!(i8, ReadData::B1, BasicReadData::B1);
read_data_from_impl!(u16, ReadData::B2, BasicReadData::B2);
read_data_from_impl!(i16, ReadData::B2, BasicReadData::B2);
read_data_from_impl!(u32, ReadData::B4, BasicReadData::B4);
read_data_from_impl!(i32, ReadData::B4, BasicReadData::B4);
read_data_from_impl!(f32, ReadData::B4, BasicReadData::B4);

impl From<bool> for BasicReadData {
    #[inline]
    fn from(val: bool) -> Self {
        BasicReadData::from(val as u8)
    }
}

/// A chunk of data to be written using SDO
///
/// ## Expedited transfer
/// * `new_data` is 1 to 4 bytes long TODO 0 possible?
/// * `offset` is 0
/// * `is_last_segment` is true
///
/// ## Segmented transfer
/// * `new_data` is 1 to 7 bytes long TODO 0 possible?
/// * `offset` is the number of bytes already written in a previous chunk
pub struct WriteData<'a> {
    pub index: u16,
    pub subindex: u8,
    pub new_data: &'a [u8],
    pub offset: usize,
    pub promised_size: Option<usize>,
    pub is_last_segment: bool,
}

pub struct BasicWriteData<'a>(&'a WriteData<'a>);

impl BasicWriteData<'_> {
    pub fn index(&self) -> u16 {
        self.0.index
    }
    pub fn subindex(&self) -> u8 {
        self.0.subindex
    }
}

/// A read-only view of [`WriteData`]
///
/// Can be directly consumed into basic objects.
/// ```
/// # use canopen::objectdictionary::datalink::WriteStream;
/// # fn demo(write_stream: WriteStream){
///  let data: u32 = write_stream.try_into().unwrap();
/// # }
///
/// ```
///
/// Can be written into a buffer.
/// ```
/// # use canopen::objectdictionary::datalink::WriteStream;
/// # fn demo(write_stream: WriteStream){
///  let mut buf = [0; 256];
///  write_stream.write_into(&mut buf).unwrap();
/// # }
/// ```
///
/// Derefs into [`WriteData`] for more advanced use cases.
pub struct WriteStream<'a>(pub(crate) &'a WriteData<'a>);

impl<'a> Deref for WriteStream<'a> {
    type Target = WriteData<'a>;

    #[inline(always)]
    fn deref(&self) -> &Self::Target {
        self.0
    }
}

impl WriteData<'_> {
    #[inline(always)]
    pub fn is_first_segment(&self) -> bool {
        self.offset == 0
    }
}

impl WriteStream<'_> {
    pub fn write_into(self, buf: &mut [u8]) -> Result<WriteStatus, ODError> {
        if let Some(promised_size) = self.promised_size {
            if promised_size > buf.len() {
                return Err(ODError::TooLong);
            }
        }

        let bytes_written = self.offset + self.new_data.len();
        if bytes_written > buf.len() {
            return Err(ODError::TooLong);
        }

        buf[self.offset..bytes_written].copy_from_slice(self.new_data);

        if self.is_last_segment {
            Ok(WriteStatus::Done { bytes_written })
        } else {
            Ok(WriteStatus::InProgress { bytes_written })
        }
    }
}

pub enum WriteStatus {
    InProgress { bytes_written: usize },
    Done { bytes_written: usize },
}

macro_rules! try_from_impl {
    ($typ:ty) => {
        impl<'a> TryFrom<BasicWriteData<'a>> for $typ {
            type Error = ODError;

            fn try_from(data: BasicWriteData<'a>) -> Result<Self, Self::Error> {
                debug_assert!(data.0.offset == 0);
                debug_assert!(data.0.is_last_segment);

                check_size(data.0.new_data.len(), core::mem::size_of::<$typ>())?;
                if let Some(size) = data.0.promised_size {
                    check_size(size, core::mem::size_of::<$typ>())?;
                }
                let data = data.0.new_data.try_into().ok().unwrap();
                Ok(<$typ>::from_le_bytes(data))
            }
        }
        impl<'a> TryFrom<WriteStream<'a>> for $typ {
            type Error = ODError;

            fn try_from(data: WriteStream<'a>) -> Result<Self, Self::Error> {
                BasicWriteData(data.0).try_into()
            }
        }
    };
}

try_from_impl!(u8);
try_from_impl!(i8);
try_from_impl!(u16);
try_from_impl!(i16);
try_from_impl!(u32);
try_from_impl!(i32);
try_from_impl!(f32);

impl<'a> TryFrom<BasicWriteData<'a>> for bool {
    type Error = ODError;

    fn try_from(data: BasicWriteData<'a>) -> Result<Self, Self::Error> {
        match u8::try_from(data)? {
            0 => Ok(false),
            1 => Ok(true),
            _ => Err(ODError::InvalidValue),
        }
    }
}

fn check_size(given: usize, expected: usize) -> Result<(), ODError> {
    match given.cmp(&expected) {
        Ordering::Less => Err(ODError::TooShort),
        Ordering::Greater => Err(ODError::TooLong),
        Ordering::Equal => Ok(()),
    }
}
