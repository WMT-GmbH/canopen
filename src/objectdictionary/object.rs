use crate::objectdictionary::OdPosition;

#[derive(Clone, Debug)]
pub struct ObjectInfo {
    pub index: u16,
    pub subindex: u8,
    pub flags: ObjectFlags,
    pub od_position: OdPosition,
}

#[derive(Clone, Copy)]
pub struct ObjectFlags(u8);

impl ObjectFlags {
    const READ_ONLY_FLAG: u8 = 0b0001_0000;
    const WRITE_ONLY_FLAG: u8 = 0b0010_0000;

    pub const fn empty() -> Self {
        ObjectFlags(0)
    }
    pub const fn set_read_only(mut self) -> Self {
        assert!(
            !self.is_write_only(),
            "Object cannot be both read-only and write-only"
        );
        self.0 |= Self::READ_ONLY_FLAG;
        self
    }
    pub const fn set_write_only(mut self) -> Self {
        assert!(
            !self.is_read_only(),
            "Object cannot be both read-only and write-only"
        );
        self.0 |= Self::WRITE_ONLY_FLAG;
        self
    }
    pub const fn set_pdo_size(mut self, size: PdoSize) -> Self {
        self.0 |= size as u8;
        self
    }

    pub const fn is_read_only(&self) -> bool {
        self.0 & Self::READ_ONLY_FLAG != 0
    }
    pub const fn is_write_only(&self) -> bool {
        self.0 & Self::WRITE_ONLY_FLAG != 0
    }
    pub const fn pdo_size(&self) -> Option<PdoSize> {
        match self.0 & 0b0000_0111 {
            0 => None,
            1 => Some(PdoSize::One),
            2 => Some(PdoSize::Two),
            4 => Some(PdoSize::Four),
            _ => unsafe { core::hint::unreachable_unchecked() }, // Other bit patterns cannot be created
        }
    }
}

impl Default for ObjectFlags {
    fn default() -> Self {
        Self::empty()
    }
}

impl core::fmt::Debug for ObjectFlags {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("ObjectFlags")
            .field("read_only", &self.is_read_only())
            .field("write_only", &self.is_write_only())
            .field("pdo_size", &self.pdo_size())
            .finish()
    }
}

#[derive(Clone, Copy, Eq, PartialEq, Debug)]
#[repr(u8)]
pub enum PdoSize {
    One = 1,
    Two = 2,
    Four = 4,
}

impl PdoSize {
    pub const fn new(val: u8) -> Option<Self> {
        match val {
            1 => Some(PdoSize::One),
            2 => Some(PdoSize::Two),
            4 => Some(PdoSize::Four),
            _ => None,
        }
    }
    pub const fn get(self) -> u8 {
        self as u8
    }
}
