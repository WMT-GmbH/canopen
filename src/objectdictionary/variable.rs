use crate::objectdictionary::datalink::{AtomicDataLink, DataLink};
use crate::objectdictionary::odcell::ODCell;
use core::num::NonZeroU8;

// TODO remove
#[derive(Clone)]
pub struct Variable<'a> {
    pub index: u16,
    pub subindex: u8,
    pub pdo_size: Option<NonZeroU8>,
    pub data: CANOpenData<'a>,
}

impl<'a> Variable<'a> {
    #[inline]
    pub fn new<T: Into<CANOpenData<'a>>>(index: u16, subindex: u8, data: T) -> Self {
        Variable {
            index,
            subindex,
            pdo_size: None,
            data: data.into(),
        }
    }
    #[inline]
    pub const fn new_datalink_cell(
        index: u16,
        subindex: u8,
        data: &'a ODCell<dyn DataLink>,
    ) -> Self {
        Variable {
            index,
            subindex,
            pdo_size: None,
            data: CANOpenData::DataLinkCell(data),
        }
    }
    #[inline]
    pub const fn new_datalink_ref(
        index: u16,
        subindex: u8,
        data: &'a dyn AtomicDataLink,
        pdo_size: Option<NonZeroU8>,
    ) -> Self {
        Variable {
            index,
            subindex,
            pdo_size,
            data: CANOpenData::DataLinkRef(data),
        }
    }
}

// TODO remove
#[derive(Clone)]
pub enum CANOpenData<'a> {
    B1([u8; 1]),
    B2([u8; 2]),
    B4([u8; 4]),
    Bytes(&'a [u8]),
    DataLinkRef(&'a dyn AtomicDataLink),
    DataLinkCell(&'a ODCell<dyn DataLink>),
    ResourceNotAvailable,
}

macro_rules! from_impl {
    ($typ:ty, $variant:path) => {
        impl From<$typ> for CANOpenData<'_> {
            #[inline]
            fn from(val: $typ) -> Self {
                $variant(val.to_le_bytes())
            }
        }
    };
}

from_impl!(u8, CANOpenData::B1);
from_impl!(i8, CANOpenData::B1);
from_impl!(u16, CANOpenData::B2);
from_impl!(i16, CANOpenData::B2);
from_impl!(u32, CANOpenData::B4);
from_impl!(i32, CANOpenData::B4);

impl From<bool> for CANOpenData<'_> {
    #[inline]
    fn from(val: bool) -> Self {
        CANOpenData::B1([val as u8])
    }
}

impl<'a> From<&'a [u8]> for CANOpenData<'a> {
    #[inline]
    fn from(val: &'a [u8]) -> Self {
        CANOpenData::Bytes(val)
    }
}

impl<'a> From<&'a str> for CANOpenData<'a> {
    #[inline]
    fn from(val: &'a str) -> Self {
        CANOpenData::Bytes(val.as_bytes())
    }
}
