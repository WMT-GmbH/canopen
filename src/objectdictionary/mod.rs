use crate::meta::{from_raw_parts_mut, DynMetadata};
use crate::objectdictionary::datalink::{
    BasicData, BasicReadData, BasicWriteData, DataLink, ReadData,
};

use crate::sdo::SDOAbortCode;

pub mod datalink;
pub mod object;
pub mod od_cell;

use crate::objectdictionary::object::{ObjectFlags, ObjectInfo};

/// # Derive macro for `OdData`
///
/// ## Example
/// ```
/// use canopen::objectdictionary::OdData;
///
/// #[derive(Default, OdData)]
/// struct Data {
///     #[canopen(index = 0x2000)]
///     number: u8
/// }
///
/// let od = Data::default().into_od();
/// ```
pub use canopen_derive::OdData;

pub trait OdData {
    type OdType;

    fn into_od(self) -> Self::OdType;
}

pub struct ObjectDictionary<T, const N: usize> {
    indices: [u16; N],
    subindices: [u8; N],
    pdo_sizes: [ObjectFlags; N],
    offsets: [usize; N],
    vtables: [DynMetadata<dyn DataLink>; N],
    pub data: T,
}

impl<T, const N: usize> ObjectDictionary<T, N> {
    #[doc(hidden)]
    pub unsafe fn new(
        indices: [u16; N],
        subindices: [u8; N],
        pdo_sizes: [ObjectFlags; N],
        offsets: [usize; N],
        vtables: [DynMetadata<dyn DataLink>; N],
        data: T,
    ) -> Self {
        ObjectDictionary {
            indices,
            subindices,
            pdo_sizes,
            offsets,
            vtables,
            data,
        }
    }

    pub fn find(&mut self, index: u16, subindex: u8) -> Result<&mut dyn DataLink, ODError> {
        let position = self.search(index, subindex)?;
        Ok(self.get(position))
    }

    pub fn read(&mut self, index: u16, subindex: u8) -> Result<ReadData, ODError> {
        self.find(index, subindex)?.read(index, subindex)
    }

    pub(crate) fn get(&mut self, position: OdPosition) -> &mut dyn DataLink {
        let mut data_ptr = &mut self.data as *mut T as *mut ();
        unsafe {
            data_ptr = data_ptr.byte_add(self.offsets[position.0]);
        }
        let metadata = self.vtables[position.0];
        let fat_ptr = from_raw_parts_mut(data_ptr, metadata);
        unsafe { &mut *fat_ptr }
    }

    pub(crate) fn get_plus(&mut self, position: OdPosition) -> (&mut dyn DataLink, OdInfo) {
        let mut data_ptr = &mut self.data as *mut T as *mut ();
        unsafe {
            data_ptr = data_ptr.byte_add(self.offsets[position.0]);
        }
        let metadata = self.vtables[position.0];
        let fat_ptr = from_raw_parts_mut(data_ptr, metadata);
        (
            unsafe { &mut *fat_ptr },
            OdInfo {
                indices: &self.indices,
                subindices: &self.subindices,
                object_flags: &self.pdo_sizes,
            },
        )
    }

    pub(crate) fn search(&self, index: u16, subindex: u8) -> Result<OdPosition, ODError> {
        od_search(&self.indices, &self.subindices, index, subindex)
    }
}

#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub struct OdPosition(pub usize);

pub struct OdInfo<'a> {
    indices: &'a [u16],
    subindices: &'a [u8],
    object_flags: &'a [ObjectFlags],
}

impl OdInfo<'_> {
    pub fn get(&self, position: OdPosition) -> ObjectInfo {
        ObjectInfo {
            index: self.indices[position.0],
            subindex: self.subindices[position.0],
            flags: self.object_flags[position.0],
            od_position: position,
        }
    }

    pub fn find(&self, index: u16, subindex: u8) -> Option<ObjectInfo> {
        let position = od_search(self.indices, self.subindices, index, subindex).ok()?;
        Some(self.get(position))
    }
}

pub(crate) fn od_search(
    indices: &[u16],
    subindices: &[u8],
    index: u16,
    subindex: u8,
) -> Result<OdPosition, ODError> {
    let mut partion_point = indices.partition_point(|&i| i < index);
    if indices.get(partion_point).copied() != Some(index) {
        return Err(ODError::ObjectDoesNotExist);
    }
    loop {
        if indices.get(partion_point).copied() != Some(index) {
            break Err(ODError::SubindexDoesNotExist);
        }
        if subindices[partion_point] == subindex {
            break Ok(OdPosition(partion_point));
        }
        partion_point += 1;
    }
}

pub struct OdArray<T, const N: usize> {
    pub array: [T; N],
}

impl<T, const N: usize> OdArray<T, N> {
    pub fn new(array: [T; N]) -> Self {
        OdArray { array }
    }
}

impl<T: BasicData, const N: usize> BasicData for OdArray<T, N> {
    fn read(&mut self, index: u16, subindex: u8) -> Result<BasicReadData, ODError> {
        if subindex == 0 {
            assert!(N <= u8::MAX as usize); // TODO inline const
            Ok((N as u8).into())
        } else if let Some(object) = self.array.get_mut(subindex as usize - 1) {
            object.read(index, subindex)
        } else {
            Err(ODError::SubindexDoesNotExist)
        }
    }

    fn write(&mut self, data: BasicWriteData, od_info: OdInfo) -> Result<(), ODError> {
        if data.subindex() == 0 {
            Err(ODError::ReadOnlyError)
        } else if let Some(object) = self.array.get_mut(data.subindex() as usize - 1) {
            object.write(data, od_info)
        } else {
            Err(ODError::SubindexDoesNotExist)
        }
    }
}

#[repr(u32)]
#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum ODError {
    ObjectDoesNotExist = SDOAbortCode::ObjectDoesNotExist as u32,
    OutOfMemory = SDOAbortCode::OutOfMemory as u32,
    UnsupportedAccess = SDOAbortCode::UnsupportedAccess as u32,
    WriteOnlyError = SDOAbortCode::WriteOnlyError as u32,
    ReadOnlyError = SDOAbortCode::ReadOnlyError as u32,
    ObjectCannotBeMapped = SDOAbortCode::ObjectCannotBeMapped as u32,
    PDOOverflow = SDOAbortCode::PDOOverflow as u32,
    ParameterIncompatibility = SDOAbortCode::ParameterIncompatibility as u32,
    InternalIncompatibility = SDOAbortCode::InternalIncompatibility as u32,
    HardwareError = SDOAbortCode::HardwareError as u32,
    WrongLength = SDOAbortCode::WrongLength as u32,
    TooLong = SDOAbortCode::TooLong as u32,
    TooShort = SDOAbortCode::TooShort as u32,
    SubindexDoesNotExist = SDOAbortCode::SubindexDoesNotExist as u32,
    InvalidValue = SDOAbortCode::InvalidValue as u32,
    ValueTooHigh = SDOAbortCode::ValueTooHigh as u32,
    ValueTooLow = SDOAbortCode::ValueTooLow as u32,
    MaxLessThanMin = SDOAbortCode::MaxLessThanMin as u32,
    ResourceNotAvailable = SDOAbortCode::ResourceNotAvailable as u32,
    GeneralError = SDOAbortCode::GeneralError as u32,
    TransferOrStorageError = SDOAbortCode::TransferOrStorageError as u32,
    LocalControlError = SDOAbortCode::LocalControlError as u32,
    DeviceStateError = SDOAbortCode::DeviceStateError as u32,
    DictionaryError = SDOAbortCode::DictionaryError as u32,
    NoDataAvailable = SDOAbortCode::NoDataAvailable as u32,
}
