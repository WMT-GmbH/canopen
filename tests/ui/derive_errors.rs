use canopen::objectdictionary::OdData;

#[derive(OdData)]
struct Test {
    a: u8,
    #[canopen()]
    b: u8,
    #[canopen(index = 0xFFFF_FFFF)]
    c: u8,
    #[canopen(index = 1, read_only, write_only)]
    d: u8,
}

#[derive(OdData)]
struct Test2 {
    #[canopen(index = 1)]
    a1: u8,
    #[canopen(index = 1)]
    a2: u8,
    #[canopen(index = 3)]
    c1: u8,
    #[canopen(index = 2)]
    b: u8,
    #[canopen(index = 3)]
    c2: u8,
}

fn main() {}
