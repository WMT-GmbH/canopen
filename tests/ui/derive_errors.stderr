error: field must have at least one #[canopen()] attribute
 --> tests/ui/derive_errors.rs:5:5
  |
5 |     a: u8,
  |     ^^^^^

error: Missing field `index`
 --> tests/ui/derive_errors.rs:6:5
  |
6 |     #[canopen()]
  |     ^^^^^^^^^^^^

error: number too large to fit in target type
 --> tests/ui/derive_errors.rs:8:23
  |
8 |     #[canopen(index = 0xFFFF_FFFF)]
  |                       ^^^^^^^^^^^

error: Object cannot be both read-only and write-only
  --> tests/ui/derive_errors.rs:10:5
   |
10 |     #[canopen(index = 1, read_only, write_only)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Duplicate index and subindex combination
  --> tests/ui/derive_errors.rs:17:5
   |
17 |     a1: u8,
   |     ^^

error: Duplicate index and subindex combination
  --> tests/ui/derive_errors.rs:21:5
   |
21 |     c1: u8,
   |     ^^
